pkg/network/common/egressip.go

```diff
@@ -290,6 +300,130 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
 	eit.syncEgressIPs()
 }
 
+func nodeIsReady(node *v1.Node) bool {
+	nodeReady := true
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == v1.NodeReady {
+			if cond.Status == v1.ConditionFalse || cond.Status == v1.ConditionUnknown {
+				nodeReady = false
+			}
+		}
+	}
+	return nodeReady
+}
+
+func (eit *EgressIPTracker) SetMonitorNodes(monitorNodes map[string]*NodeEgress) {
+	eit.monitorNodes = monitorNodes
+	if len(monitorNodes) > 0 {
+		if eit.stop == nil {
+			eit.stop = make(chan struct{})
+			go utilwait.PollUntil(DefaultPollInterval, eit.poll, eit.stop)
+		}
+	} else {
+		if eit.stop != nil {
+			close(eit.stop)
+			eit.stop = nil
+		}
+	}
+}
```

Comment:
This really feels like it needs a mutex... or if not, there need to be comments explaining why it doesn't...

(applies to all of the newly-public API...)

Reply:
It does not need a mutex, the lock on `EgressIPTracker` will already be held once this function is accessed since this function is called from `syncEgressNodeState`, which in turn is called from `syncEgressIPs` which is always called with the `EgressIPTracker` lock held. I will however add a comment mentioning this, and mention that we always expect the caller to hold the lock

Reply:
Oh, sorry, it wasn't clear from the Github review which function your comment was referencing. I now see your comment mainly referenced `SetMonitorNodes`. This function is only called from `maybeDoUpdateEgressCIDRs` (sdn-controller) which already has its own lock held: `egressIPManager` lock. We can thus assume that all sdn-controller go-routines will access this function safely. The only concern I had when going over this in my mind was w.r.t `check`, which does not hold the `EgressIPTracker` lock, it however ends up calling `SetNodeOffline` which does...but I think it might be safer if I move the locking from `SetNodeOffline` to `check`

Reply:
Circling back on my latest comment, moving the locking from `SetNodeOffline` to `check` will obviously mean that we lock much more often (every 5 seconds) with a possibility of locking for an extended period of time (depending on how many nodes we need to ping and the time it takes to process the answer). I'm not sure we want to do that? Maybe this is the reason we never did so to begin with?

Reply:
> Maybe this is the reason we never did so to begin with?

If the code doesn't hold a lock now then either (a) the current design of the code ensures that it doesn't need to hold a lock there, or (b) it's buggy and we just haven't gotten bitten yet. If the new design of the code doesn't ensure that it doesn't need to hold a lock there (eg, because now it depends on what the manager is doing, which is outside of the tracker's control), then we need to hold a lock, or else rearrange the code.

If `SetMonitorNodes` is being called in one thread by the manager and `check` is iterating `monitorNodes` in another thread, then we need a lock. One possibility would be to have `check` call a `getMonitorNodes` function that holds the lock and then gets the list of nodes then drops the lock. You'd need to make sure the handling of the individual EgressNodes was thread-safe after that.

Actually... it seems like `SetMonitorNodes` is currently wrong; it needs to merge the caller-provided node data in with the current data, to ensure that the private state for each node is kept